vm/src/stdlib/itertools.rs

```diff
@@ -488,6 +490,196 @@ mod decl {
+                    loop {
+                        let (value, new_key) = zelf.advance(vm)?;
+                        if !vm.bool_eq(new_key.clone(), old_key.clone())? {
+                            break (value, new_key);
+                        }
+                    }
+                } else {
+                    zelf.advance(vm)?
+                };
+
+                state = zelf.state.lock();
+                state.current_value = Some(value);
+                state.current_key = Some(key);
+            }
+
+            state.next_group = false;
+
+            let grouper = PyItertoolsGrouper {
+                groupby: zelf.clone(),
+            }
+            .into_ref(vm);
+
+            state.grouper = Some(Arc::downgrade(&grouper.clone().into_typed_pyobj()));
+            Ok((state.current_key.clone().unwrap(), grouper))
+        }
+
+        #[pymethod(name = "__iter__")]
+        fn iter(zelf: PyRef<Self>) -> PyRef<Self> {
+            zelf
+        }
+
+        pub(super) fn advance(&self, vm: &VirtualMachine) -> PyResult<(PyObjectRef, PyObjectRef)> {
+            let new_value = call_next(vm, &self.iterable)?;
+            let new_key = if let Some(ref kf) = self.key_func {
+                vm.invoke(kf, new_value.clone())?
+            } else {
+                new_value.clone()
+            };
+            Ok((new_value, new_key))
+        }
+    }
+
+    #[pyclass(name = "_grouper")]
+    #[derive(Debug)]
+    struct PyItertoolsGrouper {
+        groupby: PyRef<PyItertoolsGroupBy>,
+    }
+
+    type PyItertoolsGrouperRef = PyRef<PyItertoolsGrouper>;
+
+    impl PyValue for PyItertoolsGrouper {
+        fn class(vm: &VirtualMachine) -> PyClassRef {
+            vm.class("itertools", "_grouper")
+        }
+    }
+
+    #[pyimpl]
+    impl PyItertoolsGrouper {
+        #[pymethod(name = "__next__")]
+        fn next(zelf: PyRef<Self>, vm: &VirtualMachine) -> PyResult {
+            let mut state = zelf.groupby.state.lock();
+
+            if !state.is_current(&zelf) {
+                return Err(new_stop_iteration(vm));
+            }
+
+            if let Some(val) = state.current_value.clone() {
+                state.current_value = None;
+                Ok(val)
+            } else {
+                let old_key = state.current_key.clone().unwrap();
+                // Shouldn't mutex over call to advance as that executes arbitrary Python code
+                drop(state);
+                let (value, key) = zelf.groupby.advance(vm)?;
+                if vm.bool_eq(key.clone(), old_key)? {
+                    Ok(value)
+                } else {
+                    let mut state = zelf.groupby.state.lock();
+                    state.current_value = Some(value);
+                    state.current_key = Some(key);
+                    state.next_group = true;
+                    state.grouper = None;
+                    Err(new_stop_iteration(vm))
+                }
+            }
```

Comment:
This is a question because I am not used to rust convention of lock lifetime.

I want to know which one more preferred of explicit lock dropping or scope control like:
```rust
            let old_key = {
                let mut state = zelf.groupby.state.lock();

                if !state.is_current(&zelf) {
                    return Err(new_stop_iteration(vm));
                }

                if let Some(val) = state.current_value.clone() {
                    state.current_value = None;
                    return Ok(val);
                }
                
                // Shouldn't mutex over call to advance as that executes arbitrary Python code
                state.current_key.clone().unwrap();
            };

            let (value, key) = zelf.groupby.advance(vm)?;
            if vm.bool_eq(key.clone(), old_key)? {
                Ok(value)
            } else {
                let mut state = zelf.groupby.state.lock();
                state.current_value = Some(value);
                state.current_key = Some(key);
                state.next_group = true;
                state.grouper = None;
                Err(new_stop_iteration(vm))
            }
```

Reply:
That does seem a lot nicer.